calicoctl/commands/checksystem.go

```diff
@@ -0,0 +1,257 @@
+
+	// File path to Builtin kernel modules
+	modulesBuiltinPath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersionStr)
+
+	// Keep track of modules that are not found
+	modulesNotFound := []string{}
+
+	// Go through all the required modules and check Loadable and Builtin in order
+	for _, v := range requiredModules {
+		err = checkModule(modulesLoadablePath, v, kernelVersionStr)
+
+		// If not found in Loadable then check Builtin modules
+		if err != nil {
+			err = checkModule(modulesBuiltinPath, v, kernelVersionStr)
+
+			// If not found in Builtin either, add the module to the `modulesNotFound` list for later
+			if err != nil {
+				log.Printf("WARNING: Unable to detect the %s module as Loaded or Builtin module\n", v)
+				modulesNotFound = append(modulesNotFound, v)
+			}
+		}
+	}
+
+	numModulesNotFound := len(modulesNotFound)
+
+	// modsLeft keeps track of modules still not found even after checking in `lsmod`
+	// it is necessory to have a second variable to hold the list because, deleting
+	// an element in slice and assigning it to the same slice doesn't work as expected
+	// in Go as underlying slice changes while in loop messes with the sequence and it skips elements
+	modsLeft := []string{}
+	copy(modsLeft, modulesNotFound)
+
+	// Check with `lsmod` if there are still any modules in `modulesNotFound` list
+	if numModulesNotFound != 0 {
+		lsmodOut, err := exec.Command("lsmod").Output()
+		if err != nil {
+			log.Println(err)
+		}
+
+		for _, v := range modulesNotFound {
+			regex, err := regexp.Compile(v)
+			if err != nil {
+				log.Fatalf("Error: %v\n", err)
+			}
+
+			if regex.MatchString(string(lsmodOut)) {
+
+				// remove it from `modsLeft` list if found with `lsmod`
+				findAndRemove(modsLeft, v)
+				numModulesNotFound--
+			} else {
+				log.Printf("WARNING: Unable to detect the %s module with lsmod\n", v)
+			}
+		}
+	}
+
+	for _, v := range requiredModules {
+		if !contains(modulesNotFound, v) {
+			fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\t\u2713\n")
+		} else {
+			fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\t\u2717\n")
+		}
+	}
+
+	// If there are still any modules not found then return an error
+	if numModulesNotFound > 0 {
+		return errors.New("One of more kernel modules missing")
+	}
+
+	return nil
+
+}
+
+// findAndRemove finds the element passed to it and removes it from the slice
+func findAndRemove(slice []string, element string) {
+	for i, v := range slice {
+		if v == element {
+			slice = append(slice[:i], slice[i+1:]...)
+		}
+	}
+}
+
+// contains is a utility function to check if a slice contains a string
+func contains(slice []string, str string) bool {
+	for _, a := range slice {
+		if a == str {
+			return true
+		}
+	}
+	return false
+}
+
+// checkModule is a utility function used by `checkKernelModules`
+// it opens the file provided and checks if the module passed in
+// as an argument exists for the provided kernelVersion
+func checkModule(filename, module, kernelVersion string) error {
```

Comment:
Would it be easier / shorter for this to just be a grep call and check the return code?